swift/extractor/invocation/SwiftDiagnosticsConsumer.cpp

```diff
@@ -29,3 +33,29 @@ std::string SwiftDiagnosticsConsumer::getDiagMessage(swift::SourceManager& sourc
   swift::DiagnosticEngine::formatDiagnosticText(out, diagInfo.FormatString, diagInfo.FormatArgs);
   return text.str().str();
 }
+
+void SwiftDiagnosticsConsumer::forwardToLog(swift::SourceManager& sourceManager,
+                                            const swift::DiagnosticInfo& diagInfo,
+                                            const std::string& message) {
+  auto file = sourceManager.getDisplayNameForLoc(diagInfo.Loc);
+  auto [line, column] = sourceManager.getLineAndColumnInBuffer(diagInfo.Loc);
+  using Kind = swift::DiagnosticKind;
+  switch (diagInfo.Kind) {
+    case Kind::Error:
+      LOG_ERROR("{}:{}:{} {}", file, line, column, message);
+      break;
+    case Kind::Warning:
+      LOG_WARNING("{}:{}:{} {}", file, line, column, message);
+      break;
+    case Kind::Remark:
+      LOG_INFO("{}:{}:{} {}", file, line, column, message);
+      break;
+    case Kind::Note:
+      LOG_DEBUG("{}:{}:{} {}", file, line, column, message);
+      break;
+    default:
+      LOG_ERROR("unknown diagnostic kind {}, {}:{}:{} {}", diagInfo.Kind, file, line, column,
+                message);
+      break;
+  }
```

Comment:
There is a potential foot gun lurking here: one might be tempted to call `LOG_IMPL` directly with something like `LOG_IMPL(getLoggingLevelFromDiag(diagInfo.Kind), ...)`: that will happily compile, but because of how binlog works, the severity will be fixed to the first encountered one (severity is saved into a static variable).

I'm considering coding `LOG_IMPL` in a way that will give a compiler error if called with a non-constant severity, something like assigning the provided log level to a constexpr dummy variable.

Reply:
I've done that in the other PR: https://github.com/github/codeql/tree/6ef908807651851c06dabe579e68e7e088c10cd3